sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -404,69 +404,57 @@ object ColumnPruning extends Rule[LogicalPlan] {
-      })
-
-      // We only collapse these two Projects if their overlapped expressions are all
-      // deterministic.
-      val hasNondeterministic = projectList1.exists(_.collect {
-        case a: Attribute if aliasMap.contains(a) => aliasMap(a).child
-      }.exists(!_.deterministic))
-
-      if (hasNondeterministic) {
+    case p @ Project(_, agg: Aggregate) =>
+      if (hasNondeterministic(p.projectList, agg.aggregateExpressions)) {
         p
       } else {
-        // Substitute any attributes that are produced by the child projection, so that we safely
-        // eliminate it.
-        // e.g., 'SELECT c + 1 FROM (SELECT a + b AS C ...' produces 'SELECT a + b + 1 ...'
-        // TODO: Fix TransformBase to avoid the cast below.
-        val substitutedProjection = projectList1.map(_.transform {
-          case a: Attribute => aliasMap.getOrElse(a, a)
-        }).asInstanceOf[Seq[NamedExpression]]
-        // collapse 2 projects may introduce unnecessary Aliases, trim them here.
-        val cleanedProjection = substitutedProjection.map(p =>
-          CleanupAliases.trimNonTopLevelAliases(p).asInstanceOf[NamedExpression]
-        )
-        agg.copy(aggregateExpressions = cleanedProjection)
+        agg.copy(aggregateExpressions = buildCleanedProjectList(
+          p.projectList, agg.aggregateExpressions))
       }
   }
+
+  private def buildCleanedProjectList(
+      projectList1: Seq[NamedExpression],
+      projectList2: Seq[NamedExpression]): Seq[NamedExpression] = {
+    // Create a map of Aliases to their values from the child projection.
+    // e.g., 'SELECT ... FROM (SELECT a + b AS c, d ...)' produces Map(c -> Alias(a + b, c)).
+    val aliasMap = AttributeMap(projectList2.collect {
+      case a: Alias => (a.toAttribute, a)
+    })
+
+    // Substitute any attributes that are produced by the child projection, so that we safely
+    // eliminate it.
+    // e.g., 'SELECT c + 1 FROM (SELECT a + b AS C ...' produces 'SELECT a + b + 1 ...'
+    val substitutedProjection = projectList1.map(_.transform {
+      case a: Attribute => aliasMap.getOrElse(a, a)
+    })
+    // collapse 2 projects may introduce unnecessary Aliases, trim them here.
+    substitutedProjection.map(p =>
+      CleanupAliases.trimNonTopLevelAliases(p).asInstanceOf[NamedExpression]
+    )
+  }
+
+  private def hasNondeterministic(
+      projectList1: Seq[NamedExpression],
+      projectList2: Seq[NamedExpression]): Boolean = {
+    // Create a map of Aliases to their values from the child projection.
+    // e.g., 'SELECT ... FROM (SELECT a + b AS c, d ...)' produces Map(c -> Alias(a + b, c)).
+    val aliasMap = AttributeMap(projectList2.collect {
+      case a: Alias => (a.toAttribute, a)
+    })
+
+    // We only collapse these two Projects if their overlapped expressions are all
+    // deterministic.
+    projectList1.exists(_.collect {
+      case a: Attribute if aliasMap.contains(a) => aliasMap(a).child
+    }.exists(!_.deterministic))
+  }
```

Comment:
Further refactored your code a little bit:

```scala
def collectAliases(projectList: Seq[NamedExpression]): AttributeMap[Alias] = {
  AttributeMap(projectList.collect {
    case a: Alias => a.toAttribute -> a
  })
}

def haveCommonNonDeterministicOutput(
    upper: Seq[NamedExpression], lower: Seq[NamedExpression]): Boolean = {
  val aliases = collectAliases(lower)
  upper.exists(_.collect {
    case a: Attribute if aliases.contains(a) => aliases(a).child
  }).exists(!_.deterministic)
}

def buildCleanedProjectList(
    upper: Seq[NamedExpression],
    lower: Seq[NamedExpression]): Seq[NamedExpression] = {
  val aliases = collectAliases(lower)

  val rewrittenUpper = upper.map(_.transform {
    case a: Attribute => aliases.getOrElse(a, a)
  })

  rewrittenUpper.map { p =>
    CleanupAliases.trimNonTopLevelAliases(p).asInstanceOf[NamedExpression]
  }
}
```

And those inline comments need some rewording as they are now moved to different contexts.